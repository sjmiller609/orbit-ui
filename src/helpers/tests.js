import { BrowserRouter } from 'react-router-dom'
import { shape } from 'prop-types'
import { mount } from 'enzyme'

// Instantiate router context
const router = {
  history: new BrowserRouter().history,
  route: {
    location: {},
    match: {},
  },
}

const createContext = () => ({
  context: { router },
  childContextTypes: { router: shape({}) },
})

// wrap with router context - to mount RR links correctly
export function mountWrap(node) {
  return mount(node, createContext())
}
