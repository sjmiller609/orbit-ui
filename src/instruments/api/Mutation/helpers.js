const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
