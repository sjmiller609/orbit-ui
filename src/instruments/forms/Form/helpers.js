// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
