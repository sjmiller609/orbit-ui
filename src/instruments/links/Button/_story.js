import React from 'react'
import { storiesOf } from '@storybook/react'
import { Button } from 'instruments'

storiesOf('Instruments|Links.Button', module)
  .add('Default', () => <Button>Mock Button</Button>)
  .add('Small', () => <Button style="small">Mock Button</Button>)
  .add('Disabled', () => <Button disabled={true}>Mock Button</Button>)
  .add('Red Orange', () => <Button style="red-orange">Mock Button</Button>)
  .add('Red', () => <Button style="red">Mock Button</Button>)
  .add('Purple', () => <Button style="purple">Mock Button</Button>)
  .add('Beige', () => <Button style="beige">Mock Button</Button>)
  .add('Blue', () => <Button style="blue">Mock Button</Button>)
  .add('Outline', () => <Button style="outline">Mock Button</Button>)
  .add('Blue Outline', () => <Button style="blueOutline">Mock Button</Button>)
  .add('White Outline', () => <Button style="whiteOutline">Mock Button</Button>)
  .add('Arrow', () => <Button style="arrow">Mock Button</Button>)
