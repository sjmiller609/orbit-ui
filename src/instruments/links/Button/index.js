'use strict'
import PropTypes from 'prop-types'
import React from 'react'
import { Link } from 'instruments'
import s from './styles.scss'
import classnames from 'classnames'

// pass in 'to' prop for a link instead of button
const Button = ({
  onClick,
  style,
  children,
  submit,
  disabled,
  to,
  className,
  ...props
}) => {
  if (to) {
    return (
      <Link
        {...props}
        to={to}
        className={classnames(
          s.button,
          disabled ? s.disabled : null,
          s[style],
          className
        )}
        onClick={
          /* istanbul ignore next */ e => (disabled ? e.preventDefault() : e)
        }>
        {children}
      </Link>
    )
  }
  // NOTE: icons dont get added to buttons
  return (
    <Link>
      <button
        className={classnames(
          s.button,
          disabled ? s.disabled : null,
          s[style],
          className
        )}
        type={submit ? 'submit' : 'button'}
        onClick={onClick}
        disabled={disabled}>
        {children}
      </button>
    </Link>
  )
}

Button.propTypes = {
  onClick: PropTypes.func,
  style: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  submit: PropTypes.bool,
  disabled: PropTypes.bool,
  to: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  className: PropTypes.string,
}

export default Button
