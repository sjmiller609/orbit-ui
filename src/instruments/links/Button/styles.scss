@import '../../../styles/mixins.scss';

// could replace all buttons and mixins with styles here
a.button {
  display: inline-block;
}
button.button,
a.button {
  @include filmo;
  text-transform: uppercase;
  letter-spacing: 2px;
  border-radius: 3px;
  padding: 1.4em 1.8em 1.05em;
  font-size: 0.83rem;
  margin: 0;
  color: $white;
  @include shadow;
  line-height: 1.6em;
  background-image: linear-gradient(to right, $green, $blue);

  &.red-orange {
    background-image: linear-gradient(to right, $red, $orange);
  }
  &.red {
    background-image: linear-gradient(to right, $red, rgb(241, 90, 36));
  }

  &.purple {
    background-color: $purple;
    background-image: none;
  }

  &.blue {
    background-image: linear-gradient(to right, $blue, $lightBlue);
  }

  &.disabled {
    background: $beige;
    background-image: none;
    opacity: 0.6;
  }

  // outline overwrites disabled styles, must be placed after
  &.outline {
    box-shadow: inset 0 0 0 2px $beige;
    background: none;
    background-image: none;
    color: $beige;
  }
  &.blueOutline {
    box-shadow: inset 0 0 0 2px $blue;
    background: none;
    background-image: none;
    color: $blue;
  }
  &.whiteOutline {
    box-shadow: inset 0 0 0 2px $white;
    background: none;
    background-image: none;
    color: $white;
  }

  &:visited {
    color: white;
  }

  &.nomargin {
    margin: 0;
  }

  &.small {
    font-size: 0.6em;
    padding: 1.5em 3em 1.2em 3em;
  }

  .arrow {
    display: inline-block;
    min-width: 1.8rem;
    position: relative;
    span {
      position: absolute;
      font-size: 2.4rem;
      top: -1rem;
      left: 0.5rem;
      width: 1.6rem;
      overflow: hidden;
      transform: rotate(180deg);
    }
  }
}
