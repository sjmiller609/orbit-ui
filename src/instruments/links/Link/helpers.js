// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
