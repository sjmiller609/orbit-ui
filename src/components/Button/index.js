import React from 'react'
import {Link} from 'react-router-dom'
import classnames from 'classnames'

import styles from './styles.module.css'

const Button = ({ type, children, ...props }) => {
  const classes = classnames(
    styles.button,
    styles[props.theme],
    props.className,
    {
      [styles.disabled]: props.disabled
    }
  );

  if (type === 'submit') {
    return (
      <input type={type} value={props.value} className={classes} />
    )
  }

  if (type === 'link') {
    return (
      <Link {...props} className={classes}>
        {children}
      </Link>
    )
  }

  return (
    <a {...props} className={classes}>
      {children}
    </a>
  )
}

export default Button
