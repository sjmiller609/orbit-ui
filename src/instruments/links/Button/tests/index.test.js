import React from 'react'
import { mountWrap } from 'helpers/tests'
import Button from '../index'

describe('Button', () => {
  it('should be defined', () => {
    expect(Button).toBeDefined()
  })

  it('should render correctly', () => {
    const wrapper = mountWrap(
      <Button>
        <p>Children</p>
      </Button>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should render to correctly', () => {
    const wrapper = mountWrap(
      <Button to="/">
        <p>Children</p>
      </Button>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should not preventDefault', () => {
    const wrapper = mountWrap(
      <Button to="/" disabled={false}>
        <p>Children</p>
      </Button>
    )
    const event = { preventDefault: () => {} }
    jest.spyOn(event, 'preventDefault')
    wrapper.simulate('click', event)
    expect(event.preventDefault).not.toBeCalled()
  })

  it('should preventDefault if disabled', () => {
    const wrapper = mountWrap(
      <Button to="/" disabled={true}>
        <p>Children</p>
      </Button>
    )
    const event = { preventDefault: () => {} }
    jest.spyOn(event, 'preventDefault')
    wrapper.simulate('click', event)
    expect(event.preventDefault).toBeCalled()
  })

  it('should handle full height', () => {
    const wrapper = mountWrap(
      <Button disabled={true}>
        <p>Children</p>
      </Button>
    )
    const hasClass = wrapper.find('button').hasClass('disabled')
    expect(hasClass).toBeTruthy()
  })
})
