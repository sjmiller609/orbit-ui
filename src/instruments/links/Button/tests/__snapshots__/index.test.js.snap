// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<Button>
  <Link>
    <A
      className="link"
    >
      <a
        className="link"
      >
        <button
          className="button undefined"
          type="button"
        >
          <p>
            Children
          </p>
        </button>
      </a>
    </A>
  </Link>
</Button>
`;

exports[`Button should render to correctly 1`] = `
<Button
  to="/"
>
  <Link
    className="button undefined"
    onClick={[Function]}
    to="/"
  >
    <Link
      className="link button undefined"
      onClick={[Function]}
      replace={false}
      to="/"
    >
      <a
        className="link button undefined"
        href="/"
        onClick={[Function]}
      >
        <p>
          Children
        </p>
      </a>
    </Link>
  </Link>
</Button>
`;
